### PR TITLE
reject username:password in URL as per spec update

### PR DESCRIPTION
**Pull request by mattking17**
As per the spec update https://github.com/whatwg/fetch/issues/26

### CODE REVIEW

**Comment by dgraham** [comment]
Has this made it into the [Fetch specification](https://fetch.spec.whatwg.org) yet?

**Comment by mattking17** [comment]
Yes: https://fetch.spec.whatwg.org/#request-class
> If parsedURL includes credentials, throw a TypeError.

**Comment by mislav** [comment]
Does the Fetch spec include any way of performing Basic Auth then, apart from manually constructing the `Authorization` header?

**Comment by mattking17** [comment]
Also of note is inconsistent browser behavior with `username:password` in the URL (e.g. IE doesn't support it, at all).

It doesn't appear that you can set auth info in fetch other than using headers, though a simple function to produce the headers using `btoa`, or using a hardcoded string should do it.

**Comment by mattking17** [comment]
Anything else I can do here? It does follow spec so not sure why we wouldn't include it.

**Comment by mislav** [comment]
Patch looks :+1:, thanks for submitting.

This isn't a huge priority to fix so I'll let @dgraham merge at his convenience.

File: fetch.js [hunk 1]
```diff
@@ -217,6 +217,11 @@
 
   function Request(url, options) {
     options = options || {}
+
+    if (url && url.match(/^\w+?:\/\/\w+?:\w+?@/)) {
+      throw new TypeError('Credentials not allowed in request URL')
+    }
+
```
Comment (dgraham) [inline fetch.js]: Using a regular expression to detect a little-used, edge case makes me worry about false positives. We should rely on the native `URL` class to handle parsing, and add a link to the [polyfill](https://github.com/components/url) in the README.md file for old browsers.

```javascript
parsed = new URL(url)
if (parsed.username || parsed.password) {
  throw new TypeError(...)
}
```

Comment (josh) [inline fetch.js]: Depending on `URL` seems legit to me.

**Comment by mattking17** [comment]
It seems like more hoops to jump through to add another polyfill to support URL for like you said, a small edge case. Additionally, when I try to run tests in phantomjs, the `URL` object does not exist. Maybe there is a better way to detect username and password?

File: fetch.js (hunk 1 above)
Comment (mislav) [inline fetch.js]: It's a bummer that we need to pull in a dependency just to be able to raise exceptions in rare cases. I think that a regex such as this would be just fine:

```js
/^([\w-]+:)?\/\/[^\/]+@/
```

**Comment by mattking17** [comment]
Updated with suggested regex by @mislav

File: fetch.js [hunk 2]
```diff
@@ -217,6 +217,11 @@
 
   function Request(url, options) {
     options = options || {}
+
+    if (url && url.match(/^([\w-]+:)?\/\/[^\/]+@/)) {
```
Comment (mislav) [inline fetch.js]: Looks alright now. Note that I'm not advocating use of regex to parse URLs in generally. For instance in cases when we had to detect cross-origin URLs in other libraries, using URL (or at least `HTMLAnchorElement#href` and related properties) was adamant to security (/cc @mastahyeti). However in this case I think it's relatively trivial to detect whether username or password have been specified, and raise an error appropriately. The worst thing that can happen is that this check fails and the XHR request made, but with user/password ignored because XHR never reads them from the URL string.

Comment (mastahyeti) [inline fetch.js]: > However in this case I think it's relatively trivial to detect whether username or password have been specified, and raise an error appropriately.

:smile: I disagree. This regex wouldn't match

- URLs with leading whitespace
- URLs with `+`, or `.` in the scheme.
- URLs with more than two `/` after the `:`
- Other valid, but pathological URLs.

Using `URL` would be much more robust.

> The worst thing that can happen is that this check fails and the XHR request made, but with user/password ignored because XHR never reads them from the URL string.

I'm not super familiar with the behavior of Fetch. If this isn't security sensitive, then a regex might be good enough.

Comment (mislav) [inline fetch.js]: > I'm not super familiar with the behavior of Fetch. If this isn't security sensitive, then a regex might be good enough.

Thanks for jumping in. Yes, this isn't security-sensitive. If you try to use a URL that contains username/password with XHR, the username/password gets ignored because XHR wants you to pass them as arguments to `xhr.open()`.

The Fetch spec contemplated supporting username/password in URL strings but dropped that idea because of the complexities around cross-origin redirects. So, this PR adds a helpful TypeError exception so you know what's goin on. If the URL string is super-weird and the regex fails, you simply won't get a helpful exception (and no auth credentials will be sent).

**Comment by dgraham** [comment]
Closing this as stale for now. The extra error handling and spec compliance is appreciated. I'm just not comfortable matching URLs with a regular expression, and adding a dependency on the URL polyfill for IE makes using fetch in projects more difficult.
